[mlir][EmitC] Add func, call and return operations and conversions

marbre opened the pull request:
This adds a `func`, `call` and `return` operation to the EmitC dialect,
closely related to the corresponding operations of the Func dialect. In
contrast to the operations of the Func dialect, the EmitC operations do
not support multiple results. The `emitc.func` op features a
`specifiers` argument that for example allows, with corresponding
support in the emitter, to emit `inline static` functions.

Furthermore, this adds patterns and a pass to convert the Func dialect
to EmitC. A `func.func` op that is `private` is converted to
`emitc.func` with a `"static"` specifier.

marbre committed:
[mlir][EmitC] Add func, call and return operations

This adds a `func`, `call` and `return` operation to the EmitC dialect,
closely related to the corresponding operations of the Func dialect. In
contrast to the operations of the Func dialect, the EmitC operations do
not support multiple results. The `emitc.func` op features a
`specifiers` argument that for example allows, with corresponding
support in the emitter, to emit `inline static` functions.

marbre committed:
[mlir][emitc] Add `func` to `emitc` conversion

This adds patterns and a pass to convert the Func dialect to EmitC.
A `func.func` op that is `private` is converted to `emitc.func` with a
`"static"` specifier.

marbre committed:
[mlir][bazel] Add config for FuncToEmitC

marbre commented:
@AlexDenisov FYI

AlexDenisov commented:
@marbre thank you so much for working on this 🙌 

I just gave it a try and it's working great!

One minor suggestion: it would be nice to expose `populateFuncToEmitcPatterns` to combine it with other conversions without having to run a pass separately.

marbre commented:
> @marbre thank you so much for working on this 🙌
> 
> I just gave it a try and it's working great!

Thanks for testing!

> One minor suggestion: it would be nice to expose `populateFuncToEmitcPatterns` to combine it with other conversions without having to run a pass separately.

Sure! I already had this on my list, but unfortunately I didn't get to it on Friday before I had to sign off. Just split the former `FuncToEmitC.{cpp,h}` into `FuncToEmitC.{cpp,h}` and `FuncToEmitCPass.{cpp,h}`. Let me know if you have further requests or suggestions.

simon-camp commented on mlir/lib/Conversion/FuncToEmitC/FuncToEmitC.cpp:
```diff
@@ -0,0 +1,119 @@
+  LogicalResult
+  matchAndRewrite(func::CallOp callOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // multiple results func was not converted to spirv.func
```
spirv -> emitc

simon-camp commented on mlir/lib/Conversion/FuncToEmitC/FuncToEmitC.cpp:
```diff
@@ -0,0 +1,119 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements patterns to convert the Func dialect to the EmitC
```
This comment suggests that all of the func dialect operations are supported. This is not the case, as support for the `constant` and `call_indirect` operations is missing.

marbre replied:
I would read this as *This file implements patterns to convert **all operations of** the Func dialect to the EmitC* but I am open to suggestions.

simon-camp replied:
I don't have strong feelings about this, so I'm fine with keeping it as is.

simon-camp commented on mlir/lib/Conversion/FuncToEmitC/FuncToEmitC.cpp:
```diff
@@ -0,0 +1,119 @@
+    if (funcOp.isPrivate()) {
+      StringAttr specifier = rewriter.getStringAttr("static");
+      ArrayAttr specifiers = rewriter.getArrayAttr(specifier);
+      newFuncOp.setSpecifiersAttr(specifiers);
```
There is `rewriter.getStrArrayAttr`. So something like `rewriter.getStrArrayAttr({"static"})` should work.

simon-camp reviewed (approved):
LGTM. In the future we might want to make the validation of function types more ergonomic if these come up in other places, for example for the conversion of SSA values with function type, like `func.constant`.

marbre committed:
Move handling indention (NFC)

Instead of handling indention partly in `printOperation` for functions
and `printFunctionBody` this is noe fully moved to `printFunctionBody`.

jpienaar commented on mlir/lib/Conversion/FuncToEmitC/FuncToEmitC.cpp:
```diff
@@ -0,0 +1,118 @@
+    // Multiple results func was not converted to `emitc.func`.
+    if (callOp.getNumResults() > 1)
+      return rewriter.notifyMatchFailure(
+          callOp, "Only functions with zero or one result can be converted");
```
s/O/o/ (lets follow the error convention for notifications too)

jpienaar commented on mlir/lib/Conversion/FuncToEmitC/FuncToEmitC.cpp:
```diff
@@ -0,0 +1,118 @@
+
+    if (funcOp.getFunctionType().getNumResults() > 1)
+      return rewriter.notifyMatchFailure(
+          funcOp, "Only functions with zero or one result can be converted");
```
Same as above (I'll not mark the others)

jpienaar commented on mlir/lib/Conversion/FuncToEmitC/FuncToEmitCPass.cpp:
```diff
@@ -0,0 +1,49 @@
+  ConversionTarget target(getContext());
+
+  target.addLegalDialect<emitc::EmitCDialect>();
+  target.addIllegalOp<func::CallOp>();
```
I think you could do addIllegalOp<func::CallOp, func::FuncOp, ReturnOp>();

jpienaar commented on mlir/lib/Conversion/FuncToEmitC/FuncToEmitCPass.cpp:
```diff
@@ -0,0 +1,49 @@
+  target.addIllegalOp<func::ReturnOp>();
+
+  RewritePatternSet patterns(&getContext());
+  populateFuncToEmitCPatterns(patterns);
```
Could we move this to the initialize class member?

marbre replied:
Technically yes, but it seems uncommon. Most conversation passes call a function to populate the patterns within `runOnOperation()`. The only pass I am aware of which does this in the constructor is the TOSA validation pass: https://github.com/llvm/llvm-project/blob/121a0ef0bbc1a69760b69391b387e2efda0ad922/mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp#L98-L100

Since it seems not common in the code base, I would leave it as is.

jpienaar commented on mlir/lib/Dialect/EmitC/IR/EmitC.cpp:
```diff
@@ -346,6 +349,185 @@ LogicalResult ForOp::verifyRegions() {
+/// provided (leaving them alone if no entry is present). Replaces references
+/// to cloned sub-values with the corresponding value that is copied, and adds
+/// those mappings to the mapper.
+FuncOp FuncOp::clone(IRMapping &mapper) {
```
Are these tested in a flow here? I'm assuming these are close to copies from func dialect ones.

marbre replied:
Not yet and these *were* indeed copies from the Func dialect. I removed the clone methods for now and we can implement (and test) those if wer later have a use for this functionality.

jpienaar reviewed (approved):
Looks good, thanks!

simon-camp commented on mlir/lib/Dialect/EmitC/IR/EmitC.cpp:
```diff
@@ -346,6 +349,185 @@ LogicalResult ForOp::verifyRegions() {
 
   return success();
 }
+//===----------------------------------------------------------------------===//
```
Missing newline

joker-eph commented on mlir/include/mlir/Conversion/FuncToEmitC/FuncToEmitCPass.h:
```diff
@@ -0,0 +1,24 @@
+#define GEN_PASS_DECL_FUNCTOEMITC
+#include "mlir/Conversion/Passes.h.inc"
+
+std::unique_ptr<Pass> createConvertFuncToEmitC();
```
This shouldn't be needed?

marbre replied:
It seems common to most  passes to declare it in the header, but yes, it isn't needed if one uses the correct definition (`GEN_PASS_DECL_CONVERTFUNCTOEMITC` instead of `GEN_PASS_DECL_FUNCTOEMITC`). Thanks for the pointer!

marbre committed:
Drop clone methods

marbre committed:
Address review comments

marbre committed:
Disallow empty function bodies

marbre committed:
Bring ops into the right order
